File: 1 - Algorithmic Toolbox/Week_2/5 - Fibonacci Number Again/fibonacci_number_again.py
"""
Given two integers n and m, output F(n) mod m (that is, the remainder of F(n) when divided by m).

Name: Mahi Nuthanapati
Date: 06/24/2024
"""

def pisano_number( modulo: int ) -> list[int]:
    """
    Function to calculate the pisano number for a given modulo.
    
    Args:
        modulo (int): The modulo for which the pisano number is to be calculated.

    Returns:
        list[int]: The pisano number for the given modulo.
    """
    num_elements: int = ( modulo ** 2 ) - 1
    pisano_list: list[int] = []

    pisano_list.append(0)
    pisano_list.append(1)

    for i in range( 2, num_elements ):
        pisano_list.append( ( pisano_list[i - 1] + pisano_list[i - 2] ) % modulo )
        if( pisano_list[i] == 1 and pisano_list[i - 1] == 0 ):
            return pisano_list[0 : i - 1]
    
    return pisano_list

def fibonacci( fibonacci_index: int, pisano_modulo: int ) -> int:
    """
    Function to calculate the fibonacci number at the given index modulo.

    Args:
        fibonacci_index (int): index of the fibonacci number.
        pisano_modulo (int): argument for pisano_number function.
    
    Returns:
        fibonacci value at the given index modulo.
    """
    pisano_list: list[int] = pisano_number( pisano_modulo )
    num_elements_pisano: int = len( pisano_list )

    simple_fibonacci_index: int = fibonacci_index % num_elements_pisano

    return pisano_list[simple_fibonacci_index]

def main( ) -> None:
    """
    Reads user input and prints the fibonacci number at the given index modulo.
    
    Args:

    Returns:

    """
    fibonacci_index, pisano_modulo = map( int, input( ).split( ) )
    print( fibonacci( fibonacci_index, pisano_modulo ) )

if __name__ == '__main__':
    main( )

File: 1 - Algorithmic Toolbox/2 - Algorithmic Warmup/5 - Fibonacci Number Again/fibonacci_number_again.c
#include <stdio.h>
#include <stdlib.h>

long* pisano_number(long modulo, long* pisano_size)
{
    long num_elements = (modulo * modulo) - 1 ;
    long* pisano_list = (long*)malloc(num_elements * sizeof(long)) ;
    
    pisano_list[0] = 0 ; 
    pisano_list[1] = 1 ;

    for (long i = 2; i < num_elements; i++) {
        pisano_list[i] = (pisano_list[i - 1] + pisano_list[i - 2]) % modulo ;
        if (pisano_list[i] == 1 && pisano_list[i - 1] == 0) {
            *pisano_size = i - 1 ;
            return pisano_list ;
        }
    }

    *pisano_size = num_elements ;
    return pisano_list ;
}

long fibonacci(long fibonacci_index, long modulo)
{
    long pisano_size ;
    long* pisano_list = pisano_number(modulo, &pisano_size) ;

    long simple_fibonacci_index = fibonacci_index % pisano_size ;

    long result = pisano_list[simple_fibonacci_index] ;
    free(pisano_list) ; 

    return result ;
}

int main(void)
{
    long fibonacci_index ;
    long pisano_modulo ;
    scanf("%ld %ld", &fibonacci_index, &pisano_modulo) ;

    printf("%ld\n", fibonacci(fibonacci_index, pisano_modulo)) ;

    return 0 ;
}